src/Components/WebAssembly/WebAssembly/src/HotReload/WebAssemblyHotReload.cs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.HotReload;
+using Microsoft.AspNetCore.Components.WebAssembly.Services;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.HotReload
+{
+    /// <summary>
+    /// Contains methods called by interop. Intended for framework use only, not supported for use in application
+    /// code.
+    /// </summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static class WebAssemblyHotReload
+    {
+        private static readonly ConcurrentDictionary<Guid, List<(byte[] metadataDelta, byte[] ilDelta)>> _deltas = new();
+        private static readonly ConcurrentDictionary<Assembly, Assembly> _appliedAssemblies = new();
+
+        static WebAssemblyHotReload()
+        {
+            // An ApplyDelta can be called on an assembly that has not yet been loaded. This is particularly likely
+            // when we're applying deltas on app start and child components are defined in a referenced project.
+            // To account for this, wire up AssemblyLoad
+            AppDomain.CurrentDomain.AssemblyLoad += (_, eventArgs) =>
+            {
+                var loadedAssembly = eventArgs.LoadedAssembly;
```

Comment:
Does calling ApplyUpdate take a non-zero amount of time?

Reply:
It's a halt the world event. But I haven't measured how long it takes.